### PR TITLE
PacketUtils: cleaner computeChecksum code

### PR DESCRIPTION
**Pull request by clementperon**
Hi @guebe,

You recently submit a patch to fix compute checksum on bigendian and avoid an out of bound reading.

This caught my attention as this part is a bit hard to read. 

Here is a clean rework, could you review this and confirm if it works on BigEndian arch?

Thanks

### CODE REVIEW

File: Packet++/src/PacketUtils.cpp [hunk 1]
```diff
@@ -17,30 +21,32 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
...
+		// check if there is one byte left
+		if (vec[i].len % 2) {
+			// access to the last byte using an uint8_t pointer
+			uint8_t *vecBytes = (uint8_t *)vec[i].buffer;
+			uint8_t lastByte = vecBytes[vec[i].len - 1];
 			PCPP_LOG_DEBUG("1 byte left, adding value: 0x" << std::uppercase << std::hex << lastByte);
-			localSum += lastByte;
+			// We have read the latest byte manually but this byte should be interpreted
+			// as a 0xFF on LE and a 0xFF00 on BE to have a proper checksum computation
+			localSum += le16toh(lastByte);
```
Comment (guebe) [inline Packet++/src/PacketUtils.cpp:41]: on le this is a nop
on be this swaps.
this should work.

in other words we want to have the same "format" here at line 41 in local sum, as we had in line 29.

File: Tests/Packet++Test/Tests/TcpTests.cpp [hunk 2]
```diff
@@ -304,3 +308,30 @@ PTF_TEST_CASE(TcpChecksumInvalidRead)
...
+	vec[3].len = 2;
+
+	uint16_t c = pcpp::computeChecksum(vec, 3);
+	// computeChecksum return in network byte order
+	PTF_ASSERT_EQUAL(c, htons(checksum_expected));
+
+	// Adding the checksum should be equal to 0x0
+	c = pcpp::computeChecksum(vec, 4);
+	PTF_ASSERT_EQUAL(c, 0);
+} // TcpChecksumInvalidRead
```
Comment (guebe) [inline Tests/Packet++Test/Tests/TcpTests.cpp:337]: nice test.

verfied on big endian:
`root@ATL00058:/mnt/c/devel# gcc PacketUtils.cpp
root@ATL00058:/mnt/c/devel# file a.out
a.out: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=421c75fb1259673a2f88a05c9a75bc844fed5b28, for GNU/Linux 3.2.0, not stripped
root@ATL00058:/mnt/c/devel# ./a.out
61b8
0000
root@ATL00058:/mnt/c/devel# powerpc-linux-gnu-gcc -static PacketUtils.cpp
root@ATL00058:/mnt/c/devel# file a.out
a.out: ELF 32-bit MSB executable, PowerPC or cisco 4500, version 1 (SYSV), statically linked, BuildID[sha1]=843d1a9cf062cf62d025d34a4899a07af97333fc, for GNU/Linux 3.2.0, not stripped
root@ATL00058:/mnt/c/devel# ./a.out
b861
0000`

**Review by guebe — APPROVED** [review]

File: Packet++/src/PacketUtils.cpp (hunk 1 above)
Comment (guebe) [reply Packet++/src/PacketUtils.cpp:41]: would this be easier to read?:

            uint16_t lastSum = 0;
            memcpy(&lastSum, &lastByte, 1);
            localSum += lastSum;

File: Packet++/src/PacketUtils.cpp [hunk 3]
```diff
@@ -49,14 +55,16 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
...
+	// To obtain the checksum we take the ones' complement of this result
+	uint16_t result = sum;
+	result = ~result;
 
-	PCPP_LOG_DEBUG("Calculated checksum = " << sum << ", 0x" << std::uppercase << std::hex << sum);
+	PCPP_LOG_DEBUG("Calculated checksum = " << sum << ", 0x" << std::uppercase << std::hex << result);
 
-	return ((uint16_t) sum);
+	// We return the result in Network byte order
+	return htons(result);
```
Comment (guebe) [inline Packet++/src/PacketUtils.cpp:67]: isn't there more change needed in the calling code if you change the result of this function?
I mean different output now on LE (the arch where this lib is most used on)

Comment (clementperon) [reply Packet++/src/PacketUtils.cpp:67]: the return was already returned in Network byte order:
I have removed the following line
`localSum = be16toh(localSum);`

File: Packet++/src/PacketUtils.cpp [hunk 4]
```diff
@@ -1,4 +1,8 @@
+#ifndef _WIN32
+#include <arpa/inet.h>
```
Comment (guebe) [inline Packet++/src/PacketUtils.cpp:2]: perhaps use the macros from EndianPortable.h instead?
Its already included a few lines below.

File: Tests/Packet++Test/Tests/TcpTests.cpp [hunk 5]
```diff
@@ -304,3 +308,30 @@ PTF_TEST_CASE(TcpChecksumInvalidRead)
```
Comment (guebe) [inline Tests/Packet++Test/Tests/TcpTests.cpp:308]: the test above should fail on little endian after your change - it will need another htons.

File: Packet++/src/PacketUtils.cpp (hunk 4 above)
Comment (clementperon) [reply Packet++/src/PacketUtils.cpp:2]: true

File: Tests/Packet++Test/Tests/TcpTests.cpp (hunk 5 above)
Comment (clementperon) [reply Tests/Packet++Test/Tests/TcpTests.cpp:308]: No the CI passed correctly

File: Packet++/src/PacketUtils.cpp (hunk 3 above)
Comment (guebe) [inline Packet++/src/PacketUtils.cpp:67]: not obvious in the first place - but this is a nicer way than doing it on temp sums (line 43).

File: Tests/Packet++Test/Tests/TcpTests.cpp (hunk 5 above)
Comment (guebe) [reply Tests/Packet++Test/Tests/TcpTests.cpp:304]: yeah - I saw in the meantime - I also redid the test locally - then I found out that the format did not change, the code was just rearranged ...

File: Packet++/src/PacketUtils.cpp [hunk 6]
```diff
@@ -17,30 +18,35 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
...
+		// check if there is one byte left
+		if (vec[i].len % 2) {
+			// access to the last byte using an uint8_t pointer
+			uint8_t *vecBytes = (uint8_t *)vec[i].buffer;
+			uint8_t lastByte = vecBytes[vec[i].len - 1];
 			PCPP_LOG_DEBUG("1 byte left, adding value: 0x" << std::uppercase << std::hex << lastByte);
-			localSum += lastByte;
+			// We have read the latest byte manually but this byte should be interpreted
+			// as a 0xFF on LE and a 0xFF00 on BE to have a proper checksum computation
+			uint16_t lastSum;
```
Comment (guebe) [inline Packet++/src/PacketUtils.cpp:38]:
```suggestion
			uint16_t lastSum = 0;

```

Comment (guebe) [reply Packet++/src/PacketUtils.cpp:38]: without this the second byte of lastSum has undefined content

**Review by guebe — CHANGES_REQUESTED** [review]
see comment

File: Packet++/src/PacketUtils.cpp [hunk 7]
```diff
@@ -35,7 +35,10 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
 			PCPP_LOG_DEBUG("1 byte left, adding value: 0x" << std::uppercase << std::hex << lastByte);
 			// We have read the latest byte manually but this byte should be interpreted
 			// as a 0xFF on LE and a 0xFF00 on BE to have a proper checksum computation
-			localSum += le16toh(lastByte);
+			uint16_t lastSum;
+			memcpy(&lastSum, &lastByte, 1);
```
Comment (clementperon) [inline Packet++/src/PacketUtils.cpp:39]: @seladb proposition from @guebe, IMO the le16toh() is more explicit.

Which one do you prefer?

File: Packet++/src/PacketUtils.cpp [hunk 8]
```diff
@@ -35,7 +35,10 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
 			PCPP_LOG_DEBUG("1 byte left, adding value: 0x" << std::uppercase << std::hex << lastByte);
 			// We have read the latest byte manually but this byte should be interpreted
 			// as a 0xFF on LE and a 0xFF00 on BE to have a proper checksum computation
-			localSum += le16toh(lastByte);
+			uint16_t lastSum = 0;
+			memcpy(&lastSum, &lastByte, 1);
```
Comment (clementperon) [inline Packet++/src/PacketUtils.cpp:39]: @seladb proposition from @guebe, IMO the le16toh() is more explicit.

Which one do you prefer?

Comment (guebe) [reply Packet++/src/PacketUtils.cpp:39]: for me both would be fine ...

File: Packet++/src/PacketUtils.cpp (hunk 3 above)
Comment (clementperon) [reply Packet++/src/PacketUtils.cpp:67]: maybe i'm wrong but i think the previous way was wrong as then the carry count would be wrong.

Comment (guebe) [reply Packet++/src/PacketUtils.cpp:67]: I think it was also OK. because of https://datatracker.ietf.org/doc/html/rfc1071 Page 3, (B)  Byte Order Independence

Comment (guebe) [reply Packet++/src/PacketUtils.cpp:67]: btw. why do we do the carry twice? once on the local sum and once on sum. Should be enough to do it only on sum.

please have a look at the original version of tcpdump. I very much like the simplicity of it:
https://github.com/the-tcpdump-group/tcpdump/blob/a91d20cc8491f2afcc04e51474244710481244b9/print-ip.c#L276

also the implementation and comment of the next function is notable:
https://github.com/the-tcpdump-group/tcpdump/blob/a91d20cc8491f2afcc04e51474244710481244b9/print-ip.c#L312

File: Packet++/src/PacketUtils.cpp [hunk 9]
```diff
@@ -17,30 +18,35 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
...
+			localSum += lastSum;
+
 			PCPP_LOG_DEBUG("Local sum = " << localSum << ", 0x" << std::uppercase << std::hex << localSum);
 		}
 
+		// carry count is added to the sum
 		while (localSum>>16)
 		{
 			localSum = (localSum & 0xffff) + (localSum >> 16);
 		}
```
Comment (guebe) [inline Packet++/src/PacketUtils.cpp:49]: btw. why do we do the carry twice? once on the local sum and once on sum. Should be enough to do it only on sum.

please have a look at the original version of tcpdump. I very much like the simplicity of it:
https://github.com/the-tcpdump-group/tcpdump/blob/a91d20cc8491f2afcc04e51474244710481244b9/print-ip.c#L276

also the implementation and comment of the next function is notable:
https://github.com/the-tcpdump-group/tcpdump/blob/a91d20cc8491f2afcc04e51474244710481244b9/print-ip.c#L312

Comment (guebe) [reply Packet++/src/PacketUtils.cpp:49]: I think I was wrong. Not adding the carrys of local sum will give us problems for very long messages.

File: 3rdParty/EndianPortable/include/EndianPortable.h [hunk 10]
```diff
@@ -207,6 +207,10 @@
 #define htole32(x) le32toh(x)
 #endif
 
+#ifndef _WIN32
+#include <arpa/inet.h> // for htonl()
+#endif
```
Comment (seladb) [inline 3rdParty/EndianPortable/include/EndianPortable.h:212]: I'm not sure why this change is needed, where do we use `htonl` in `computeChecksum()`?

File: Packet++/src/PacketUtils.cpp (hunk 8 above)
Comment (seladb) [reply Packet++/src/PacketUtils.cpp:39]: yes, I think it'd be cleaner to use `be16toh()`. We can assume `be` because network order is always BE

File: Packet++/src/PacketUtils.cpp (hunk 9 above)
Comment (seladb) [reply Packet++/src/PacketUtils.cpp:49]: Yes, we need to add the carry 😃

File: Packet++/src/PacketUtils.cpp [hunk 11]
```diff
@@ -49,14 +55,16 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
 	{
 		sum = (sum & 0xffff) + (sum >> 16);
 	}
-
 	PCPP_LOG_DEBUG("Sum before invert = " << sum << ", 0x" << std::uppercase << std::hex << sum);
 
-	sum = ~sum;
+	// To obtain the checksum we take the ones' complement of this result
+	uint16_t result = sum;
+	result = ~result;
```
Comment (seladb) [inline Packet++/src/PacketUtils.cpp:62]: I'm not sure why the extra `result` variable is needed. What's wrong with the previous code?

File: Packet++/src/PacketUtils.cpp (hunk 8 above)
Comment (clementperon) [reply Packet++/src/PacketUtils.cpp]: done

File: Packet++/src/PacketUtils.cpp (hunk 11 above)
Comment (clementperon) [reply Packet++/src/PacketUtils.cpp:60]: to have a proper print std::hex << result;

File: 3rdParty/EndianPortable/include/EndianPortable.h (hunk 10 above)
Comment (clementperon) [reply 3rdParty/EndianPortable/include/EndianPortable.h:212]: We use htons() i Added the comment because we use htonl() in htonll()

Comment (seladb) [reply 3rdParty/EndianPortable/include/EndianPortable.h:212]: I'm not sure I understand... where in `computeChecksum()` do we use `htonl` or htonll`? In other words - what will fail if we remove this change?

File: Packet++/src/PacketUtils.cpp [hunk 12]
```diff
@@ -49,14 +53,16 @@ uint16_t computeChecksum(ScalarBuffer<uint16_t> vec[], size_t vecSize)
...
+	// To obtain the checksum we take the ones' complement of this result
+	uint16_t result = sum;
+	result = ~result;
 
-	PCPP_LOG_DEBUG("Calculated checksum = " << sum << ", 0x" << std::uppercase << std::hex << sum);
+	PCPP_LOG_DEBUG("Calculated checksum = " << sum << ", 0x" << std::uppercase << std::hex << result);
 
-	return ((uint16_t) sum);
+	// We return the result in Network byte order
+	return htons(result);
```
Comment (seladb) [inline Packet++/src/PacketUtils.cpp:65]: Since we have `3rdParty/EndianPortable/include/EndianPortable.h` We don't need to use `htons`, `htonl` or `htonll` directly (which are sometimes platform specific). Instead we should use `htobe16()`

File: Packet++/src/PacketUtils.cpp (hunk 11 above)
Comment (seladb) [reply Packet++/src/PacketUtils.cpp:60]: oh ok, that makes sense 👍

File: 3rdParty/EndianPortable/include/EndianPortable.h (hunk 10 above)
Comment (clementperon) [reply 3rdParty/EndianPortable/include/EndianPortable.h]: i have switched htons to htobe16

File: Packet++/src/PacketUtils.cpp (hunk 12 above)
Comment (clementperon) [reply Packet++/src/PacketUtils.cpp]: done

File: 3rdParty/EndianPortable/include/EndianPortable.h [hunk 13]
```diff
@@ -61,25 +61,25 @@
...
   #define htole32(x) (x)
-  #define be32toh(x) ntohl(x)
+  #define be32toh(x) __builtin_bswap32(x)
   #define le32toh(x) (x)
 
-  #define htobe64(x) htonll(x)
+  #define htobe64(x) __builtin_bswap64(x)
   #define htole64(x) (x)
-  #define be64toh(x) ntohll(x)
+  #define be64toh(x) __builtin_bswap64(x)
```
Comment (seladb) [inline 3rdParty/EndianPortable/include/EndianPortable.h:82]: Why is this change needed? Wouldn't it be simpler to keep the previous code? That way we don't need to add a special `#ifedf` for MSVC

Comment (clementperon) [reply 3rdParty/EndianPortable/include/EndianPortable.h:82]: But htonll isn't defined and is it guarantee that network byte order is always big endian ?

Comment (clementperon) [reply 3rdParty/EndianPortable/include/EndianPortable.h:82]: If you want to keep the previous format,  the Big Endian on windows will fail as it's using GCC built-in

Comment (seladb) [reply 3rdParty/EndianPortable/include/EndianPortable.h:82]: `htonll` isn't defined in which platform? We currently don't use it so I'm not sure it matters...
Anyway I assume this is not needed for this PR, is that correct?
Maybe we can remove this change from this PR and think later if we want to add it or not. What do you think?

Comment (clementperon) [reply 3rdParty/EndianPortable/include/EndianPortable.h:82]: We don't use but C preprocessor comply that the macro doesn't exist.

Yes not related to this MR, I will drop these changes

Comment (clementperon) [reply 3rdParty/EndianPortable/include/EndianPortable.h]: @seladb done
